### ISSUE TITLE
[Epic]  Investigate alternatives to SQLite for the system catalog

### ISSUE THREAD

**Issue by heeringa** [MaterializeInc/materialize#11577]
Principle: Materialize is dependable
Initiative: Materialize stays up and running

**Problem**:  Supporting a highly available system means eliminating single points of failure that create noticeable delays in performance or user experience as well as loss of data.  Right now we use SQLite on a single node as our backing store for the system catalog, which has two problems:  it could go down and not come back up as quickly as we’d like (single point of failure) and it’s backed by the filesystem so data could more easily be corrupted.  Here we’re interested in answering how important this problem is to a highly available system.
Success:  Understanding the tradeoffs to staying with SQLite versus using other, more robust backing stores.
Priority: Medium
Time Horizon:  April
Effort: 1 Week

**Comment by benesch** [comment]
I actually think we can call this one closed, @heeringa? You've mostly enumerated the downsides of SQLite-backed metadata storage already. Here's the full enumeration that I'm aware of:

  * The durability of EBS is relatively low: 99.8% - 99.9% according to AWS. That means we _will_ observe SQLite corruption in production.
  * A given EBS volume is constrained to a single AZ. Materialize Platform will be unable to tolerate a single AZ failure with SQLite-backed metadata storage.
  * Sizing/resizing EBS volumes is difficult via Kubernetes. This will hamper scalability for customers with sizeable metadata, as we currently can't automatically resize these EBS volumes if the SQLite database starts to fill the volume.
  * We've heard anecdotally that EBS volumes can get wedged: that is, stuck in an in-between state where they cannot be attached to a new pod, and you have to wait many hours or contact AWS support.

In summary: SQLite on EBS prevents us from tolerating single AZ failure and is operationally challenging to boot. I think we are all aligned on moving metadata storage to a proper highly-available database.

**Comment by benesch** [comment]
If the work item here is rather "decide _which_ HA database to replace SQLite with", that's a different question that we're actively exploring!

**Comment by heeringa** [comment]
@benesch - the alternatives / replacements is encapsulated by this epic.  So let's keep it alive and feel free to use the space here to provide updates.

**Comment by benesch** [comment]
Ack. We're going to need longer than one week for this. I have a few hypotheses that we're exploring. We'll commit to one soon, but we'll need to re-evaluate periodically over the next 2-3 months.

**Comment by benesch** [comment]
Design doc out: https://github.com/MaterializeInc/materialize/blob/d7adebee00be8f99d5b1eecb43988928a0c8e84e/doc/developer/design/20220404_metadata_storage.md

**Comment by JLDLaughlin** [comment]
I _believe_ we can close this issue as a duplicate of https://github.com/MaterializeInc/cloud/issues/2476, but correct me if I'm wrong @benesch / @heeringa !

**Comment by benesch** [comment]
I think probably we want to consider MaterializeInc/cloud#2476 to be blocked on this epic! And then #2476 can focus on _just_ what system tables need to be added to facilitate audit logs.

**Comment by jychen7** [comment]
> e use SQLite on a single node as our backing store for the system catalog

fyi, may consider https://litestream.io as well
https://tailscale.com/blog/database-for-2022/

**Comment by heeringa** [comment]
@benesch and @JLDLaughlin - there is a design around audit logging in [#2476](https://github.com/MaterializeInc/cloud/issues/2476) - is that sufficient, now, to close this epic? 

Also, how should I think about how the audit log relates to the system catalog?

**Comment by JLDLaughlin** [comment]
yes! we've committed to Aurora RDS, although I believe we have a checkpoint in ~June to decide if it was the right decision.
